tests/EdgeToEdge/Routes/ShowDonationConfirmationRouteTest.php

```diff
@@ -148,4 +149,26 @@ public function testWhenNoDonation_accessIsDenied(): void {
 		$this->assertStringContainsString( self::ACCESS_DENIED_TEXT, $responseContent );
 	}
 
+	public function testRateLimitTimestampIsStoredInSession(): void {
+		$this->modifyEnvironment( function ( FunFunFactory $factory ): void {
+			$this->donation = $this->newStoredDonation( $factory );
+		} );
+		$client = $this->createClient();
+
+		$client->request(
+			'GET',
+			'show-donation-confirmation',
+			[
+				'id' => $this->donation->getId(),
+				'accessToken' => self::CORRECT_ACCESS_TOKEN
+			]
+		);
+
+		/** @var SessionInterface $session */
+		$session = static::getContainer()->get( 'session' );
+		$donationTimestamp = $session->get( FunFunFactory::DONATION_RATE_LIMIT_SESSION_KEY );
+		$this->assertNotNull( $donationTimestamp );
+		$this->assertEqualsWithDelta( time(), $donationTimestamp->getTimestamp(), 5.0, 'Timestamp should be not more than 5 seconds old' );
```

Comment:
If you really want to decouple the code from external services, you could use [wmde/Clock](https://github.com/wmde/Clock) as a dependency of `DonationSubmissionRateLimiter` and swap the `SystemClock` with a `StubClock` in this test. But that's not a must (because it would also make the test setup more complicated and add new setters and getters to FunFunFactory).

Reply:
I think it'll be okay as it is, it's a very basic class